### PR TITLE
added project_id to defined variables

### LINKED ISSUE #1693: Add gitlab project id to defined variable
### What feature do you want to see added?

Gitlab webhook sends post request that contains "project_id". Currently, this is not mapped to a [defined variable](https://plugins.jenkins.io/gitlab-plugin/#plugin-content-defined-variables). This can be useful when repository are trying to call gitlab's api, for example when uploading generic package. 

### Upstream changes

_No response_

### Are you interested in contributing this feature?

_No response_

### PR DESCRIPTION
**Pull request by Aditya-PS-05**
## closes #1693 

added project_id to defined variables

### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by krisstern** [comment]
Hi @Aditya-PS-05, it looks like your proposed code changes introduced some unexpected test failures. You may want to check out the JUnit testing we have and see if any adjustment to our existing tests you will need to make.

**Comment by Aditya-PS-05** [comment]
@krisstern , try to test with mvn but it's showing errors in pom.xml. Should I create a new issue? Or only it's os specific in my ubuntu22.04 LTS

**Comment by krisstern** [comment]
Hi @Aditya-PS-05, could you please share what errors you see regarding `pom.xml`? Without more info I cannot tell whether the issue is specific to your setup or not.

**Comment by krisstern** [comment]
So I have just pulled your code and ran `mvn clean verify` with it. I don't see any issues with `pom.xml` though. What Java JDK version are you using?

**Comment by Aditya-PS-05** [comment]
When running `mvn clean install -X`

errors: 
```
[DEBUG] Writing tracking file /home/aditya/.m2/repository/io/jenkins/tools/bom/bom-weekly/3613.v584fca_12cf5c/_remote.repositories
[DEBUG] Writing tracking file /home/aditya/.m2/repository/io/jenkins/tools/bom/bom-weekly/3613.v584fca_12cf5c/bom-weekly-3613.v584fca_12cf5c.pom.lastUpdated
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Unknown packaging: hpi @ org.jenkins-ci.plugins:gitlab-plugin:${revision}${changelist}, /home/aditya/my-work/open-source/gitlab-plugin/pom.xml, line 13, column 14
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
[ERROR] Unknown packaging: hpi @ org.jenkins-ci.plugins:gitlab-plugin:${revision}${changelist}, /home/aditya/my-work/open-source/gitlab-plugin/pom.xml, line 13, column 14

    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:397)
    at org.apache.maven.graph.DefaultGraphBuilder.collectProjects (DefaultGraphBuilder.java:414)
    at org.apache.maven.graph.DefaultGraphBuilder.getProjectsForMavenReactor (DefaultGraphBuilder.java:405)
    at org.apache.maven.graph.DefaultGraphBuilder.build (DefaultGraphBuilder.java:82)
    at org.apache.maven.DefaultMaven.buildGraph (DefaultMaven.java:507)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:219)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR]   
[ERROR]   The project org.jenkins-ci.plugins:gitlab-plugin:1.9.7-SNAPSHOT (/home/aditya/my-work/open-source/gitlab-plugin/pom.xml) has 1 error
[ERROR]     Unknown packaging: hpi @ org.jenkins-ci.plugins:gitlab-plugin:${revision}${changelist}, /home/aditya/my-work/open-source/gitlab-plugin/pom.xml, line 13, column 14
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
```

and when running `mvn test`

errors
```
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Unknown packaging: hpi @ org.jenkins-ci.plugins:gitlab-plugin:${revision}${changelist}, /home/aditya/my-work/open-source/gitlab-plugin/pom.xml, line 13, column 14
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.jenkins-ci.plugins:gitlab-plugin:1.9.7-SNAPSHOT (/home/aditya/my-work/open-source/gitlab-plugin/pom.xml) has 1 error
[ERROR]     Unknown packaging: hpi @ org.jenkins-ci.plugins:gitlab-plugin:${revision}${changelist}, /home/aditya/my-work/open-source/gitlab-plugin/pom.xml, line 13, column 14
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
```

**Comment by krisstern** [comment]
Regarding https://github.com/jenkinsci/gitlab-plugin/pull/1722#issuecomment-2461797075:

@Aditya-PS-05 Did you try to run `mvn clean verify`, and which version of Java JDK are you using?

**Comment by Aditya-PS-05** [comment]
> mvn clean verify

openjdk 21.0.5 2024-10-15 LTS

when running `mvn clean verify` 

res:
```
aditya@aditya-HP-Laptop-15s-du3xxx:~/my-work/open-source/gitlab-plugin$ mvn clean verify
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Unknown packaging: hpi @ org.jenkins-ci.plugins:gitlab-plugin:${revision}${changelist}, /home/aditya/my-work/open-source/gitlab-plugin/pom.xml, line 13, column 14
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.jenkins-ci.plugins:gitlab-plugin:1.9.7-SNAPSHOT (/home/aditya/my-work/open-source/gitlab-plugin/pom.xml) has 1 error
[ERROR]     Unknown packaging: hpi @ org.jenkins-ci.plugins:gitlab-plugin:${revision}${changelist}, /home/aditya/my-work/open-source/gitlab-plugin/pom.xml, line 13, column 14
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
```

**Comment by krisstern** [comment]
I have checked I don't think it is a `pom.xml` error, it is more like a local setup issue for you

**Comment by krisstern** [comment]
And I am asking this for the 3rd time: Which Java JDK are you using?

**Comment by Aditya-PS-05** [comment]
> And I am asking this for the 3rd time: Which Java JDK are you using?

Sir, I already answered it 

it's OpenJDK 21.0.5 2024-10-15 LTS

**Comment by krisstern** [comment]
My apologies you answered right after quoting `mvn clean verify` without any context so I could not tell you were specifying your Java version. Please be more careful in your replies moving forward. How about if you switch to Java 17 instead? What do you see?

**Comment by krisstern** [comment]
And which Maven version are you currently using?

**Comment by Aditya-PS-05** [comment]
> And which Maven version are you currently using?

Apache Maven 3.6.3

**Comment by Aditya-PS-05** [comment]
> My apologies you answered right after quoting `mvn clean verify` without any context so I could not tell you were specifying your Java version. Please be more careful in your replies moving forward. How about if you switch to Java 17 instead? What do you see?

got it

**Comment by krisstern** [comment]
@Aditya-PS-05 I think your Maven version is too old. The latest version is 3.9.x. So you may want to download the latest one say from here: https://maven.apache.org/download.cgi.

File: src/main/java/com/dabsquared/gitlabjenkins/cause/CauseData.java [hunk 1]
```diff
@@ -115,7 +115,7 @@ public final class CauseData {
         this.actionType = Objects.requireNonNull(actionType, "actionType must not be null.");
         this.sourceProjectId = Objects.requireNonNull(sourceProjectId, "sourceProjectId must not be null.");
         this.targetProjectId = Objects.requireNonNull(targetProjectId, "targetProjectId must not be null.");
-        this.projectId = Objects.requireNonNull(projectId, "projectId must not be null.");
+        this.projectId = projectId == null ? 0 : projectId;
```
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/cause/CauseData.java:118]:
```suggestion
        this.projectId = Objects.requireNonNull(projectId, "projectId must not be null.");
```

We should not change the original logic for handling the `projectId` in "src/main/java/com/dabsquared/gitlabjenkins/cause/CauseData.java"

**Review by krisstern — CHANGES_REQUESTED** [review]

**Pull request closed without merging**